### PR TITLE
Simplify release process

### LINKED ISSUE #58: Automate release process
Current release process is documented in the readme and is fairly painful. The most ideal process would be to push a tag and have the release happen automatically (but I am not clear what happens to the changelog in this case). Using a make command would also be acceptable if it can do a bunch of changes in a row like bump the version, publish, bump version to snapshot again, and then automatically close & release the repo from staging to maven via sonatype.

### PR DESCRIPTION
**Pull request by pavolloffay**
Supersedes #622
Closes #58 #622 

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -196,20 +201,14 @@ configure(subprojects.findAll {it.name != 'jaeger-thrift'}) {
...
-    return ext.developmentVersion
-}
-
 def getProperty(name, defaultValue) {
     project.properties[name] ?: defaultValue
 }
+
+release {
+    git {
+        requireBranch = 'master|release-*|simplify-release-create-commits'
```
Comment (pavolloffay) [inline build.gradle]: I will remove `simplify-release-create-commits` once it is approved.

File: RELEASE.md [hunk 2]
```diff
@@ -4,19 +4,20 @@
...
-     * commit with the comment "Back to development" (for example, https://github.com/jaegertracing/jaeger-client-java/commit/da9726d3ba7309947882e3c621516c70b4bc83dc)
+  1. Commit your `CHANGELOG` changes as `git commit -m "Preparing release 0.20.0" -s`
+  1. Create and push tag with the new version `git tag release-0.20.0 && git push origin release-0.20.0`. 
+  Do not update the version in `gradle.properties` manually.
+  1. The build for the `release-` tag does following:
+     * It updates the version in `gradle.properties` and creates commit and tag `v0.20.0`
+     * It updates version to the next `SNAPSHOT` and commits it
+     * The next build for the tag publishes artifact to Nexus. If the build fails it might 
+     be necessary to drop staging repositories and restart the build or just close the repository.
+  1. Create a release on GitHub for the new tag. Use the changes from the `CHANGELOG` as the description.
```
Comment (yurishkuro) [inline RELEASE.md]: creating a release on GitHub creates yet another tag?

File: travis/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+set -o errexit -o nounset -o pipefail
+
+
+if [[ "$TRAVIS_TAG" =~ ^release-[[:digit:]]+\.[[:digit:]]+\.[[:digit:]]+?$ ]]; then
+    echo "We are on release- tag"
+    echo "bumping versions and creating vx.x.x tag"
+    echo "final artifact will be published in build for the tag"
+    version=$(echo "${TRAVIS_TAG}" | sed 's/^release-//')
+    ./gradlew release release -Prelease.useAutomaticVersion=true -Prelease.releaseVersion=${version}
```
Comment (yurishkuro) [inline travis/deploy.sh]: is  this what makes the commits? Do we need to set up git identity for the bot?

Comment (pavolloffay) [reply travis/deploy.sh]: travis creates commits, we already inject GH token there.

File: RELEASE.md (hunk 2 above)
Comment (pavolloffay) [reply RELEASE.md]: Yes, the first tag triggers release and the second tag is tags the commit with final version. This is exactly how maven release works.

When you checkout the `release-20.0.0` tag the project still uses previous snapshot `0.19.0-SNAPSHOT`. When you checkout `v20.0.0` the project uses final version `20.0.0`

File: travis/deploy.sh (hunk 3 above)
Comment (pavolloffay) [reply travis/deploy.sh]: Yes the gradle plugin creates the commit, I will verify in my fork if it works with our current config. If no I will add required credentials.

File: RELEASE.md (hunk 2 above)
Comment (yurishkuro) [reply RELEASE.md]: What I meant was when you create a release on GitHub you need to provide a new tag, and by that point we'll already have release-x.y.z and vx.y.z

Comment (pavolloffay) [reply RELEASE.md]: The github release should point to `vx.y.z` as this one has a final version set.

File: .travis.yml [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
+  depth: false
 services:
 - docker
 env:
   global:
   - DOCKER_VERSION=1.11.1-0~trusty
   - DOCKER_COMPOSE_VERSION=1.7.0
   - COMMIT=${TRAVIS_COMMIT::8}
+  # GH_TOKEN
+  - secure: "U5MNSISi7mUv+oDT9wwnrv2i90CaD6mSHrkkzW9tChikMsssq8cnWO8KwQ226Qbv2V16dLHr8sq504At2dKOzsxbBxLOkH7OXEt1uCrMErrMVL82XmKkThD6KTWSiDRdYLFoNbGv/q0B+U6ZZwPdTDibVquBw71cNDUuq8GdivCyT9f/7GQNUTVHjHmFRlU3lTAx9Z5rSKW6ZyL4tVFheTNNYO758Swg2pVVXRQbO9ZOfW7Mg5OKuR1i1gtpB4gpTtcVPykE3hEPd/loOHUZSQZu91c8uSk3Oxp1Pk8vKsUseebQoyaGXxnL1LZtNpbZHz+6p9PzlAEWIWNmUme0ji1B42zXgMxY2QNNj4Y27aNj9eiOqI6oUBjN1qyErgH467DEzMsbMs/weYW9KlwiJ9OfIHvRZ+BM5uxyeBs/xSb5UBlULmecsTDI5xWYJmv70iOLmSNZ4eRjFUkge89sGi/EgJEbLKE4Z8BXRbmdIMVycsZTfirS0k85NkWgNIQ4dMBhm5V28kZaPiO/kri4FXoge/Zbvqt1FkTY/puEh+AC7cP9S+PjpdTe9vtHc78BXP0cIrS/JL/wELiQumSDqXSnnQ+1OCB8+LLAMIr2tfA5YXC44UoyyaoMSWk7RO6KyUVgYDk3tTU7SFwsyeYVBGXr8X21tMxHE0Yu4nbS/FM="
```
Comment (pavolloffay) [inline .travis.yml]: @yurishkuro I have added just `GH_TOKEN`, we do the same in OT contrib in many repos.

I have generated the token myself, we could switch to bot, but I am not sure if we have one.

**Comment by pavolloffay** [comment]
Once merged I will cut patch release to test it out.

**Comment by pavolloffay** [comment]
The 0.35.4 build https://travis-ci.org/jaegertracing/jaeger-client-java/jobs/530816415 failed but artifacts were upload to nexus and staging repository was closed.

**Pull request merged**